verification/cbmc/proofs/aws_cryptosdk_hdr_write/aws_cryptosdk_hdr_write_harness.c

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/cryptosdk/header.h>
+#include <aws/cryptosdk/private/header.h>
+#include <make_common_data_structures.h>
+#include <proof_helpers/make_common_data_structures.h>
+#include <proof_helpers/proof_allocators.h>
+#include <proof_helpers/utils.h>
+
+// A generator function as described in the comment in aws_cryptosdk_hash_elems_array_init_stub.c.
+// Also see line 33 of the Makefile.
+void array_list_item_generator(struct aws_array_list *elems) {
+    assert(elems->item_size == sizeof(struct aws_hash_element));
+    for (size_t index = 0; index < elems->length; ++index) {
+        struct aws_hash_element *val = (struct aws_hash_element *)((uint8_t *)elems->data + (elems->item_size * index));
+        // Due to the checks in aws_cryptosdk_enc_ctx_size, no string can have a length > UINT16_MAX
+        struct aws_string *key = ensure_string_is_allocated_nondet_length();
+        __CPROVER_assume(aws_string_is_valid(key));
+        __CPROVER_assume(key->len <= UINT16_MAX);
+        val->key                 = key;
+        struct aws_string *value = ensure_string_is_allocated_nondet_length();
+        __CPROVER_assume(aws_string_is_valid(value));
+        __CPROVER_assume(value->len <= UINT16_MAX);
+        val->value = value;
+    }
+}
+
+void aws_cryptosdk_hdr_write_harness() {
+    /* Nondet Input */
+    struct aws_cryptosdk_hdr *hdr = ensure_nondet_hdr_has_allocated_members(MAX_TABLE_SIZE);
+    size_t *bytes_written;
+    uint8_t *outbuf;
+    size_t outlen;
+
+    /* Assumptions */
+    __CPROVER_assume(aws_cryptosdk_hdr_members_are_bounded(hdr, MAX_EDK_LIST_ITEMS, MAX_BUFFER_SIZE));
+    __CPROVER_assume(IMPLIES(hdr != NULL, aws_byte_buf_is_bounded(&hdr->iv, MAX_IV_LEN)));
+
+    /* Precondition: The edk list has allocated list elements */
+    ensure_cryptosdk_edk_list_has_allocated_list_elements(&hdr->edk_list);
+    __CPROVER_assume(aws_cryptosdk_hdr_is_valid(hdr));
+
+    __CPROVER_assume(hdr->enc_ctx.p_impl != NULL);
+    ensure_hash_table_has_valid_destroy_functions(&hdr->enc_ctx);
+    size_t empty_slot_idx;
+    __CPROVER_assume(aws_hash_table_has_an_empty_slot(&hdr->enc_ctx, &empty_slot_idx));
```

Comment:
Why do you need this assumption and why it's not mapped to an assertion in the actual code? Could you add a comment?

Reply:
Great point! So I included this assumption as it is used in the `enc_ctx_serialize` proof and `hdr_write` calls that function. However, I've just experimented locally and it seems this assumption actually isn't needed (and could also be removed from the `enc_ctx_serialize` proof). I believe this assumption in general is used to guarantee that the loop unwinding values given to CBMC are sufficient when we are iterating over a hash_table with an aws_hash_iter. Since `enc_ctx_size`(where such a loop occurs) is stubbed out in both this proof and `enc_ctx_serialize`, this assumption isn't actually needed.
We had a discussion a while back about mapping this assumption to the source code and DSN gave a good explanation of what this function does and why we don't map it directly to the actual code (https://github.com/aws/aws-encryption-sdk-c/pull/608#discussion_r508500101).

Reply:
Actually I think I was wrong above. I've just removed the assumptions from enc_ctx_size and the proof remains unchanged. Perhaps this assumption was once needed to guarantee termination but no longer?